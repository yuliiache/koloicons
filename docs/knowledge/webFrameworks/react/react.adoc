== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







