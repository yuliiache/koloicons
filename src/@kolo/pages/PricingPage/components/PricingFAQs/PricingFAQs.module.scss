@import "@kolo/design";

.section {
  padding-top: 76px;
  padding-bottom: 118px;

  @media screen and (max-width: 1023px) {
    padding-bottom: 116px;
  }

  @media screen and (max-width: 767px) {
    padding-top: 44px;
    padding-bottom: 90px;
  }
}

.container {
  width: 100%;
  padding: 0 16px;
  @include container;
}

.title {
  margin-bottom: 44px;

  font-size: 1.6rem;
  font-weight: 900;
  text-align: center;
  color: $TitleColor;

  @media screen and (max-width: 767px) {
    margin-bottom: 24px;
    font-size: 0.9rem;
  }
}

.list {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  row-gap: 68px;

  @media screen and (max-width: 1023px) {
    row-gap: 32px;
  }

  @media screen and (max-width: 767px) {
    display: block;
  }
}
