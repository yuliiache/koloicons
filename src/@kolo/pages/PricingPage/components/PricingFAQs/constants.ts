export const PRICING_FAQS_KEY = 'pricingFAQs';
