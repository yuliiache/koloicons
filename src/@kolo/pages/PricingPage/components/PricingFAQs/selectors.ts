import {RootState} from 'app/store';
import {createSelector, Selector} from 'reselect';

import {PRICING_FAQS_KEY} from './constants';
import {FaqsState} from './slice/pricingFAQsSlice';
import {FaqsData} from './types/FaqsData';

const getPricingFAQs: Selector<RootState, FaqsState> = (state) => state[PRICING_FAQS_KEY];

export const pricingFAQsSelector: Selector<RootState, FaqsData[]> = createSelector(
  getPricingFAQs,
  (pricingFAQs: FaqsState): FaqsData[] => pricingFAQs.faqsList
);
