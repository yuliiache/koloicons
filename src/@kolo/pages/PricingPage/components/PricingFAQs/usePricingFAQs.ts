import {useAppDispatch, useAppSelector} from 'app/hooks';
import {useEffect} from 'react';

import {pricingFAQsSelector} from './selectors';
import {getPricingFAQsStartAction} from './slice/pricingFAQsSlice';
import {FaqsData} from './types/FaqsData';

interface PricingFAQsHookReturn {
  faqsList: FaqsData[];
}

const usePricingFAQs = (): PricingFAQsHookReturn => {
  const faqsList: FaqsData[] = useAppSelector(pricingFAQsSelector);
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(getPricingFAQsStartAction());
  }, []);

  return {faqsList};
};

export default usePricingFAQs;
