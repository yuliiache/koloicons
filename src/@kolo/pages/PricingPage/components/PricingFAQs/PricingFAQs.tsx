import {FC} from 'react';

import PricingFAQsItem from './components/PricingFAQsItem/PricingFAQsItem';
import styles from './PricingFAQs.module.scss';
import {FaqsData} from './types/FaqsData';
import usePricingFAQs from './usePricingFAQs';

const PricingFAQs: FC = () => {
  const {faqsList} = usePricingFAQs();

  if (!faqsList || !faqsList.length) {
    return null;
  }

  return (
    <section className={styles.section}>
      <div className={styles.container}>
        <h2 className={styles.title}>Frequently Asked Questions</h2>
        <ul className={styles.list}>
          {faqsList.map(({question, answer}: FaqsData) => (
            <PricingFAQsItem
              key={question}
              question={question}
              answer={answer}
            />
          ))}
        </ul>
      </div>
    </section>
  );
};

export default PricingFAQs;
