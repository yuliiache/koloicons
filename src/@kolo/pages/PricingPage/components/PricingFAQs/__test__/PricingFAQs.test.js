import {render} from '@testing-library/react';

import PricingFAQs from '../PricingFAQs';
import usePricingFAQs from '../usePricingFAQs';

jest.mock('../usePricingFAQs.ts');

describe('PricingFAQs component', () => {
  const mockFaqsList = [
    {
      question: 'Question 1',
      answer: 'Answer 1',
    },
    {
      question: 'Question 2',
      answer: 'Answer 2',
    },
  ];
  const usePricingFAQsHook = () => usePricingFAQs.mockReturnValue({faqsList: mockFaqsList});

  it('should render correctly with pricing FAQs data', () => {
    usePricingFAQsHook();

    const {getByText} = render(<PricingFAQs />);

    mockFaqsList.forEach(({question, answer}) => {
      expect(getByText(question)).toBeInTheDocument();
      expect(getByText(answer)).toBeInTheDocument();
    });
  });

  it('should render snapshot successfully', () => {
    usePricingFAQsHook();
    const {asFragment} = render(<PricingFAQs />);

    expect(asFragment()).toMatchSnapshot();
  });
});
