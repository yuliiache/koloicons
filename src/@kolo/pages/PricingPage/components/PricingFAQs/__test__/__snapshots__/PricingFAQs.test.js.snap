// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PricingFAQs component should render snapshot successfully 1`] = `
<DocumentFragment>
  <section
    class="section"
  >
    <div
      class="container"
    >
      <h2
        class="title"
      >
        Frequently Asked Questions
      </h2>
      <ul
        class="list"
      >
        <li
          class="item"
        >
          <p
            class="question"
          >
            Question 1
          </p>
          <p>
            Answer 1
          </p>
        </li>
        <li
          class="item"
        >
          <p
            class="question"
          >
            Question 2
          </p>
          <p>
            Answer 2
          </p>
        </li>
      </ul>
    </div>
  </section>
</DocumentFragment>
`;
